### PR TITLE
feat(mcp): validate AI-generated write payloads against schemas [RHCLOUD-47091]

### PR DESCRIPTION
**Pull request by lpichler**
## Link(s) to Jira
- https://redhat.atlassian.net/browse/RHCLOUD-47091

## Description of Intent of Change(s)

**What:** Adds a semantic validation layer for MCP write tool payloads that catches domain-specific errors before they reach Django views.

**Why:** LLMs can generate plausible-looking but subtly malformed payloads — wrong permission formats, invalid UUIDs, or overly permissive role definitions. The existing `_validate_tool_arguments()` only checks FastMCP-generated JSON schemas (structural types). This deeper layer catches semantic issues that structural validation cannot.

**How:** A new `_validate_write_payload()` function runs after JSON schema validation but before write confirmation in `_handle_tools_call()`. It dispatches to tool-specific validators:

- **Permission format** (V1): validates `application:resource_type:verb` with non-empty, non-whitespace-padded segments
- **V2 permissions**: validates dict has non-empty `application`, `resource_type`, `operation`
- **UUID format**: validates all UUID arguments before view delegation using existing `is_valid_uuid()`
- **Permission count**: caps roles at 100 permissions with a clear error

Error messages are LLM-friendly — they describe what's wrong and how to fix it, enabling the model to self-correct.

Covers 24 write tools across create/update/delete operations for groups, roles, workspaces, role bindings, and cross-account requests.

## Local Testing

```bash
# Run lint
pipenv run tox -e lint

# Run new validation tests
pipenv run tox -e py312-fast -- tests.management.test_mcp_views.MCPWritePayloadValidationUnitTests tests.management.test_mcp_views.MCPWritePayloadValidationIntegrationTests

# Run all MCP tests
pipenv run tox -e py312-fast -- tests.management.test_mcp_views
```

To manually test via MCP endpoint:
```bash
# Malformed permission string → rejected with clear error
curl -X POST /_private/_a2s/mcp/ -d '{"jsonrpc":"2.0","method":"tools/call","id":1,"params":{"name":"create_role_v1","arguments":{"name":"Bad","access":[{"permission":"only-two-parts:read"}]}}}'

# Invalid UUID → rejected with clear error
curl -X POST /_private/_a2s/mcp/ -d '{"jsonrpc":"2.0","method":"tools/call","id":1,"params":{"name":"delete_role_v1","arguments":{"role_uuid":"not-a-uuid"}}}'
```

## Checklist
- [x] if API spec changes are required, is the spec updated? *(no API spec changes — internal MCP validation only)*
- [x] are there any pre/post merge actions required? if so, document here. *(none)*
- [x] are theses changes covered by unit tests? *(34 new tests: 27 unit + 7 integration)*
- [x] if warranted, are documentation changes accounted for? *(no docs changes needed — internal validation)*
- [x] does this require migration changes? *(no)*
- [ ] is there known, direct impact to dependent teams/components?

## Secure Coding Practices Checklist Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Practices Checklist
- [x] Input Validation *(core of this PR — validates all write tool inputs)*
- [x] Output Encoding
- [x] Authentication and Password Management
- [x] Session Management
- [x] Access Control
- [x] Cryptographic Practices
- [x] Error Handling and Logging *(LLM-friendly errors, proper logging)*
- [x] Data Protection
- [x] Communication Security
- [x] System Configuration
- [x] Database Security
- [x] File Management
- [x] Memory Management
- [x] General Coding Practices

## Summary by Sourcery

Add semantic validation for MCP write tool payloads to catch domain-specific issues before executing write operations.

Bug Fixes:
- Reject malformed permission strings, invalid UUIDs, and overly large permission sets for MCP write tools with clear error responses.

Enhancements:
- Introduce centralized write payload validation covering V1 and V2 role, workspace, role binding, and cross-account request tools, integrated into MCP tools/call handling.
- Improve error messaging for invalid MCP write payloads to be more descriptive and suitable for LLM-driven correction.

Tests:
- Add unit and integration tests verifying write payload validation helpers and end-to-end MCP endpoint behavior for valid and invalid write requests.

### CODE REVIEW

File: rbac/management/mcp_views.py [hunk 1]
```diff
@@ -5434,6 +5435,175 @@ def _validate_tool_arguments(tool_name: str, arguments: dict[str, Any]) -> str |
...
+    if not is_valid_uuid(value):
+        return f"'{field_name}' value '{value}' is not a valid UUID."
+    return None
+
+
+def _validate_write_payload(tool_name: str, arguments: dict[str, Any], request: HttpRequest) -> str | None:
+    """Validate semantic constraints on write tool arguments.
+
+    Runs after JSON schema validation to catch domain-specific issues:
+    - Permission strings match 'application:resource_type:verb' format
```
Comment (lpichler) [inline rbac/management/mcp_views.py]: **low**: Unused `request` parameter in `_validate_write_payload`

The `request` parameter is passed to `_validate_write_payload()` but never used within the function body. It may have been included for future extensibility but creates unnecessary coupling.

File: rbac/management/mcp_views.py [hunk 2]
```diff
@@ -5434,6 +5435,175 @@ def _validate_tool_arguments(tool_name: str, arguments: dict[str, Any]) -> str |
...
+    if tool_name == "patch_role_v1":
+        return _validate_uuid_field(arguments.get("role_uuid", ""), "role_uuid")
+
+    if tool_name == "add_roles_to_group":
+        for i, role_id in enumerate(arguments.get("roles") or []):
+            error = _validate_uuid_field(role_id, f"roles[{i}]")
+            if error:
+                return error
+        return None
+
```
Comment (lpichler) [inline rbac/management/mcp_views.py]: **low**: `add_roles_to_group` validates role UUIDs but not `group_uuid`

The `add_roles_to_group` handler validates each entry in `roles` as a UUID but does not validate `group_uuid` if provided. Other handlers validate their primary UUID argument.

File: rbac/management/mcp_views.py [hunk 3]
```diff
@@ -5434,6 +5435,175 @@ def _validate_tool_arguments(tool_name: str, arguments: dict[str, Any]) -> str |
     return None
 
 
+# --- Semantic write payload validation ---
+
```
Comment (lpichler) [inline rbac/management/mcp_views.py:5623]: **informational**: Permission cap constant is hardcoded

_MAX_PERMISSIONS_PER_ROLE = 100 is a hardcoded constant. This is an MCP-specific guardrail and a reasonable default, but is not configurable per deployment.

**Comment by lpichler** [comment]
## Code Review Findings

**File:** `rbac/management/mcp_views.py` (line 1)

**informational**: Eight write tools have no semantic validation rules

15 of 23 write tools have explicit validation. The 8 uncovered tools (add_principals_to_group, create_cross_account_request, create_group, create_workspace, delete_group, remove_principals_from_group, remove_roles_from_group, update_group) return None (pass). Most lack UUID params needing validation, so this is defensible.

File: rbac/management/mcp_views.py [hunk 4]
```diff
@@ -58,6 +58,7 @@
 from management.role_binding.model import RoleBinding, RoleBindingGroup, RoleBindingPrincipal
 from management.role_binding.view import RoleBindingViewSet
 from management.tenant_mapping.v2_activation import is_v2_write_activated
+from management.utils import is_valid_uuid
 from management.workspace.view import WorkspaceViewSet
```
Comment (lpichler) [inline rbac/management/mcp_views.py:64]: **medium**: High cyclomatic complexity in _validate_write_payload function due to excessive conditional branching

The function has 15+ conditional branches (lines 70-183) handling different tool_name cases in a large if-elif chain. Each tool validation follows similar patterns (UUID validation, permission validation) but is implemented inline.

File: rbac/management/mcp_views.py [hunk 5]
```diff
@@ -5434,6 +5435,175 @@ def _validate_tool_arguments(tool_name: str, arguments: dict[str, Any]) -> str |
     return None
```
Comment (lpichler) [inline rbac/management/mcp_views.py:5621]: **medium**: Monolithic file growth: _validate_write_payload and helpers add 170 lines to a 6032-line file that already contains tool definitions, JSON-RPC transport, view dispatch, confirmation tokens, metrics, and now semantic validation. mcp_views.py has 124 function definitions and is growing into a god module. The new validation functions (_validate_permission_format, _validate_v2_permission, _validate_uuid_field, _validate_write_payload) are self-contained and have no dependency on anything else in mcp_views.py except the is_valid_uuid import from utils. They should be extracted to a dedicated module (e.g., management/mcp_validation.py) to maintain separation of concerns and keep the file navigable.

mcp_views.py is 6032 lines with 124 function definitions. The new validation block (lines 5437-5609 in the diff) is purely functional with no coupling to other code in this file -- it only imports is_valid_uuid from management.utils.

Comment (lpichler) [inline rbac/management/mcp_views.py:5621]: **medium**: Duplicated validation logic: The permission format validation in _validate_permission_format duplicates the same 3-part colon-split check that already exists in AccessSerializer.validate_permission (role/serializer.py line 159). The V2 permission field validation in _validate_v2_permission partially duplicates checks in RoleV2Service._validate_and_resolve_permissions (role/v2_service.py line 93). Since MCP tools delegate to these same views via _call_view_write, the serializers will perform these checks again when the request reaches the view layer. This creates a maintenance risk where validation rules may drift between the two locations.

AccessSerializer.validate_permission (serializer.py:159-168) checks len(split_value) != 3, same as _validate_permission_format. RoleV2Service._validate_and_resolve_permissions (v2_service.py:93) validates permission dicts, same as _validate_v2_permission. Both paths execute because MCP tools call _call_view_write which goes through the serializer/service layer.

File: rbac/management/mcp_views.py [hunk 6]
```diff
@@ -5434,6 +5435,175 @@ def _validate_tool_arguments(tool_name: str, arguments: dict[str, Any]) -> str |
...
+    - UUID arguments are valid UUID format
+    - Permission counts don't exceed limits
+    """
+    if tool_name in ("create_role_v1", "update_role_v1"):
+        if tool_name == "update_role_v1":
+            error = _validate_uuid_field(arguments.get("role_uuid", ""), "role_uuid")
+            if error:
+                return error
+        access_list = arguments.get("access") or []
+        if len(access_list) > _MAX_PERMISSIONS_PER_ROLE:
```
Comment (lpichler) [inline rbac/management/mcp_views.py]: **medium**: Tool-name-based dispatch via long if/elif chain: _validate_write_payload uses 12 sequential if-return blocks keyed on tool name strings. This pattern is brittle -- adding a new write tool requires remembering to add a validation case here, and the coupling between tool registration (via @register_tool) and validation rules is implicit. The existing _TOOL_CONFIG registry pattern could be extended with a 'validator' callable field to co-locate validation with tool configuration, making it impossible to forget validation for new write tools.

The function contains 12 separate if blocks checking tool_name against string literals: 'create_role_v1', 'update_role_v1', 'create_role', 'update_role', 'patch_role_v1', 'add_roles_to_group', 'create_role_bindings', 'delete_role_v1', 'bulk_delete_roles', 'update_workspace', 'move_workspace', 'delete_workspace', 'update_role_binding', 'update_cross_account_request', 'patch_cross_account_request'. ToolConfig at line 223 is the existing registry dataclass that could host a validator field.

Comment (lpichler) [inline rbac/management/mcp_views.py]: **low**: Unused parameter: _validate_write_payload accepts request: HttpRequest as its third parameter but never references it in the function body. The parameter is passed from _handle_tools_call at line 5930 and every test creates a request object just to pass it. This is either dead code from a previous design or forward-looking for future tenant-aware validation, but in its current form it violates interface cleanliness and adds unnecessary test boilerplate.

The function signature is def _validate_write_payload(tool_name: str, arguments: dict[str, Any], request: HttpRequest) -> str | None. Searching the function body (lines 5497-5609 in the source) reveals no reference to 'request'. Every unit test creates a RequestFactory request and sets request.tenant solely to pass this unused parameter.

File: rbac/management/mcp_views.py [hunk 7]
```diff
@@ -5434,6 +5435,175 @@ def _validate_tool_arguments(tool_name: str, arguments: dict[str, Any]) -> str |
...
+        return f"'{field_name}' value '{value}' is not a valid UUID."
+    return None
+
+
+def _validate_write_payload(tool_name: str, arguments: dict[str, Any], request: HttpRequest) -> str | None:
+    """Validate semantic constraints on write tool arguments.
+
+    Runs after JSON schema validation to catch domain-specific issues:
+    - Permission strings match 'application:resource_type:verb' format
+    - UUID arguments are valid UUID format
```
Comment (lpichler) [inline rbac/management/mcp_views.py]: **low**: User-supplied values reflected verbatim in error messages and log output without length limits. The _validate_uuid_field function includes the raw user value in its error string (e.g. "'role_uuid' value '{value}' is not a valid UUID") and _validate_permission_format includes the full permission string. Since FastMCP-generated schemas do not enforce maxLength on string fields, an attacker could supply an arbitrarily long string that gets reflected in the JSON-RPC error response and written to the logger.warning call. This could contribute to log bloat or log injection if the logging pipeline does not handle newlines/control characters. In practice, the risk is low because: (1) the endpoint requires authentication, (2) Python's logger.warning with %s formatting does not interpret format strings in the value, and (3) the response is JSON (not HTML), so XSS is not a concern.

File: rbac/management/mcp_views.py [hunk 8]
```diff
@@ -58,6 +58,7 @@
 from management.role_binding.model import RoleBinding, RoleBindingGroup, RoleBindingPrincipal
 from management.role_binding.view import RoleBindingViewSet
 from management.tenant_mapping.v2_activation import is_v2_write_activated
+from management.utils import is_valid_uuid
 from management.workspace.view import WorkspaceViewSet
 from mcp.server.fastmcp import FastMCP
 from prometheus_client import Counter, Histogram
```
Comment (lpichler) [inline rbac/management/mcp_views.py:66]: **low**: Validation function returns early on first error without validating remaining access entries, potentially hiding multiple validation issues

In _validate_write_payload for create_role_v1/update_role_v1 (lines 81-86 in diff), the function iterates through access_list and returns immediately on the first permission format error. This means if multiple permissions are malformed, only the first error is reported. The user must fix errors one at a time rather than seeing all issues at once.

File: rbac/management/mcp_views.py [hunk 9]
```diff
@@ -5758,6 +5928,12 @@ def _handle_tools_call(request: HttpRequest, request_id: Any, params: dict[str,
...
 
+    if config.write:
+        write_error = _validate_write_payload(tool_name, arguments, request)
+        if write_error:
+            logger.warning("mcp: tools/call tool='%s' write payload validation failed: %s", tool_name, write_error)
+            return _error_response(request_id, -32602, f"Invalid params for tool '{tool_name}': {write_error}")
+
     if config.write and _is_write_confirmation_enabled():
         if confirmation_token:
             valid, error_msg = _validate_confirmation_token(confirmation_token, tool_name, arguments, org_id)
```
Comment (lpichler) [inline rbac/management/mcp_views.py:6204]: **info**: Two-layer validation architecture: JSON schema validation followed by domain-specific semantic validation

The PR implements a two-phase validation pipeline: (1) `_validate_tool_arguments()` at line 5940 validates JSON schema compliance (types, required fields), then (2) `_validate_write_payload()` at line 5945+ validates semantic constraints (permission format, UUID validity, business rules like max permissions). This separates structural validation from domain logic validation, making each layer simpler and more focused.

File: rbac/management/mcp_views.py [hunk 10]
```diff
@@ -5434,6 +5435,175 @@ def _validate_tool_arguments(tool_name: str, arguments: dict[str, Any]) -> str |
...
+        if not part or part != part.strip():
+            return (
+                f"Permission '{permission}' has an empty or whitespace-padded {label}. "
+                "Each segment must be a non-empty string with no leading/trailing spaces."
+            )
+    return None
+
+
+def _validate_v2_permission(perm: dict[str, Any], index: int) -> str | None:
+    """Validate a V2 permission dict has non-empty application, resource_type, and operation."""
```
Comment (lpichler) [inline rbac/management/mcp_views.py:5646]: **info**: Defensive validation with helpful error messages for AI-generated payloads

The validation functions return detailed, actionable error messages. Example: `_validate_permission_format` (line 5422) returns 'Permission \'X\' is malformed. Expected format \'application:resource_type:verb\' (e.g. \'cost-management:cost_model:read\')' with an example. The `_MAX_PERMISSIONS_PER_ROLE` check (line 5476) suggests splitting into multiple roles. These messages guide AI agents (and humans) to fix issues without requiring code inspection.

File: rbac/management/mcp_views.py [hunk 11]
```diff
@@ -5434,6 +5435,175 @@ def _validate_tool_arguments(tool_name: str, arguments: dict[str, Any]) -> str |
     return None
```
Comment (lpichler) [inline rbac/management/mcp_views.py:5619]: **info**: Semantic validation runs AFTER schema validation but BEFORE authorization/confirmation flow

The validation order in `_handle_tools_call` is: (1) schema validation (line 5940), (2) write payload validation (line 5945), (3) confirmation token check (line 5946+). This placement catches malformed input early (before generating confirmation tokens or checking permissions), reducing wasted work and preventing confusing error messages after user confirmation.

File: tests/management/test_mcp_views.py [hunk 12]
```diff
@@ -7945,3 +7950,376 @@ def test_validate_invalid_returns_message(self):
         result = _validate_tool_arguments("hello", {"message": 999})
```
Comment (lpichler) [inline tests/management/test_mcp_views.py:8247]: **info**: Comprehensive test coverage strategy: unit tests for validators + integration tests for full pipeline

The PR adds two test classes: `MCPWritePayloadValidationUnitTests` (line 7951+) with 30+ unit tests for individual validator functions, and `MCPWritePayloadValidationIntegrationTests` (line 8044+) with integration tests that call tools through the full MCP endpoint. This dual-layer testing mirrors the dual-layer validation and ensures both components work in isolation and together.

File: rbac/management/mcp_views.py (hunk 11 above)
Comment (lpichler) [inline rbac/management/mcp_views.py:5619]: **info**: Field-specific validators return None on success, error string on failure for clean composition

All validator functions (`_validate_permission_format`, `_validate_v2_permission`, `_validate_uuid_field`) use the same signature: take input, return `None` on success or error string on failure. This allows `_validate_write_payload` to compose them with simple `if error: return error` chains (lines 5472-5485, 5490-5504, etc.). The pattern is consistent across all validators, making the code predictable.

File: rbac/management/mcp_views.py [hunk 13]
```diff
@@ -5434,6 +5435,175 @@ def _validate_tool_arguments(tool_name: str, arguments: dict[str, Any]) -> str |
...
+    """Validate a V1 permission string matches 'application:resource_type:verb' format."""
+    parts = permission.split(":")
+    if len(parts) != 3:
+        return (
+            f"Permission '{permission}' is malformed. "
+            "Expected format 'application:resource_type:verb' (e.g. 'cost-management:cost_model:read')."
+        )
+    for i, label in enumerate(("application", "resource_type", "verb")):
+        part = parts[i]
+        if not part or part != part.strip():
```
Comment (lpichler) [inline rbac/management/mcp_views.py:5637]: **info**: UUID validation reuses existing utility function instead of duplicating validation logic

The PR imports `is_valid_uuid` from `management.utils` (line 9) and wraps it in `_validate_uuid_field` (line 5453) to provide field-specific error messages. This avoids duplicating the UUID parsing logic (which uses `UUID(str(value))` try/except) and ensures consistent validation across the codebase. The wrapper adds the field name to error messages ('role_uuid' value 'X' is not a valid UUID').

File: rbac/management/mcp_views.py [hunk 14]
```diff
@@ -5434,6 +5435,175 @@ def _validate_tool_arguments(tool_name: str, arguments: dict[str, Any]) -> str |
...
+            return (
+                f"Permission '{permission}' has an empty or whitespace-padded {label}. "
```
Comment (lpichler) [inline rbac/management/mcp_views.py:5639]: **info**: The _validate_permission_format function validates structural format (3 colon-separated non-empty parts) but does not restrict the character set allowed in each segment. Any characters including shell metacharacters, SQL operators, HTML tags, or Unicode can pass through as valid permission segments. While this validation layer is purely a pre-check before the data reaches Django serializers and ORM (which provide their own sanitization), it represents a missed opportunity to reject obviously invalid content early at the MCP boundary. The existing permission model in the database likely constrains these values, but rejecting suspicious characters at this validation layer would provide defense-in-depth.

File: rbac/management/mcp_views.py (hunk 1 above)
Comment (lpichler) [inline rbac/management/mcp_views.py]: **low**: Unused `request` parameter in `_validate_write_payload`

The `request` parameter is passed to `_validate_write_payload()` but never used within the function body. It may have been included for future extensibility but creates unnecessary coupling.

File: rbac/management/mcp_views.py (hunk 2 above)
Comment (lpichler) [inline rbac/management/mcp_views.py]: **low**: `add_roles_to_group` validates role UUIDs but not `group_uuid`

The `add_roles_to_group` handler validates each entry in `roles` as a UUID but does not validate `group_uuid` if provided. Other handlers validate their primary UUID argument.

File: rbac/management/mcp_views.py (hunk 3 above)
Comment (lpichler) [inline rbac/management/mcp_views.py:5623]: **informational**: Permission cap constant is hardcoded

_MAX_PERMISSIONS_PER_ROLE = 100 is a hardcoded constant. This is an MCP-specific guardrail and a reasonable default, but is not configurable per deployment.

**Comment by lpichler** [comment]
## Code Review Findings

**File:** `tests/management/test_mcp_views.py` (line 231)

**medium**: Unit tests are isolated from integration flow - validation is called before confirmation token check in the actual flow (line 5944 vs 5945-5967), but unit tests don't verify this ordering

In _handle_tools_call, _validate_write_payload is called at line 5944 (after schema validation), before the confirmation token logic at lines 5945-5967. Unit tests in MCPWritePayloadValidationUnitTests test validation functions in isolation but don't verify they're called at the right point in the request lifecycle.



---

**File:** `rbac/management/mcp_views.py` (line 111)

**medium**: Repetitive UUID validation pattern duplicated across multiple tool handlers

UUID validation logic is repeated in multiple places: add_roles_to_group (lines 111-114), create_role_bindings (lines 117-139), bulk_delete_roles (lines 146-149), etc. Each follows identical error-checking pattern.



---

**File:** `rbac/management/mcp_views.py` (line 194)

**medium**: Write payload validation runs after confirmation_token.pop() but before confirmation flow, allowing token to be consumed even on validation failure

Line 5938: confirmation_token = arguments.pop('confirmation_token', None) happens before the new validation at line 5931 (diff line 194). If validation fails, the token has already been removed from arguments, which could break retry attempts. The validation should happen before mutating the arguments dict.



---

**File:** `tests/management/test_mcp_views.py` (line 502)

**medium**: Integration tests are missing coverage for v2 write tools (create_role, update_role) - only create_role_v1 and other v1 tools are tested in the integration layer

Integration test class has tests for create_role_v1 (line 535), delete_role_v1 (line 545), bulk_delete_roles (line 579), but only one v2 test (test_v2_permission_missing_field_rejected at line 569). The validation logic has different code paths for v1 vs v2 tools (lines 70-105 in mcp_views.py), but integration coverage heavily favors v1.



---

**File:** `tests/management/test_mcp_views.py` (line 559)

**medium**: test_valid_payload_passes_through uses negative assertions that are brittle - checking that error strings are NOT in the JSON dump of the entire response is fragile and doesn't confirm validation actually passed

Lines 565-566 use assertNotIn('malformed', json.dumps(data)) and assertNotIn('not a valid UUID', json.dumps(data)). This doesn't verify validation succeeded - it only checks error keywords aren't present. The response could fail for other reasons (auth, permission, etc.) and the test would still pass.



---

**File:** `tests/management/test_mcp_views.py` (line 326)

**low**: Repeated RequestFactory().get('/') boilerplate across test methods

Lines 327-329, 336-338, 348-350, and many other test methods repeat the same 3 lines creating a mock request with tenant. This appears ~20 times in the new test code.



---

**File:** `tests/management/test_mcp_views.py` (line 502)

**low**: Integration tests don't verify the full error response structure - they only check error code and message substring, missing validation of the complete JSON-RPC error shape

Integration tests like test_malformed_permission_rejected_via_endpoint (line 533) assert data['error']['code'] and check substring in data['error']['message'], but don't verify the error response has the complete JSON-RPC structure (e.g., that 'jsonrpc' field is present, 'id' matches request, no unexpected fields).



---

**File:** `tests/management/test_mcp_views.py` (line 341)

**low**: Unit test creates mock request objects that don't match the real request object shape - validation function receives the request parameter but doesn't use it, creating dead code path

_validate_write_payload accepts a 'request' parameter (line 62 in mcp_views.py) but never uses it in the implementation. Unit tests create RequestFactory().get('/') mocks and set request.tenant (lines 337-338, 349-350, etc.) but this is unused. Either the parameter will be needed in the future or it's leftover from refactoring.



---

**File:** `rbac/management/mcp_views.py` (line 107)

**low**: Inconsistent validation pattern - patch_role_v1 only validates role_uuid but doesn't validate permission format

Lines 107-108 show patch_role_v1 only validates role_uuid, unlike create_role_v1 and update_role_v1 (lines 70-86) which validate both UUID and permission format. If patch_role_v1 accepts permissions, this is inconsistent.



---

**File:** `rbac/management/mcp_views.py` (line 22)

**low**: Permission format validation uses string.split() in a loop for each permission during write operations

In _validate_permission_format(), permission.split(':') is called, followed by iterating over parts to check for empty/whitespace segments. For roles with many permissions (up to 100 per _MAX_PERMISSIONS_PER_ROLE), this creates temporary list objects repeatedly.



---

**File:** `tests/management/test_mcp_views.py` (line 365)

**low**: Permission count limit test uses edge case (exactly limit+1) but doesn't test the boundary (exactly at limit)

test_create_role_v1_exceeds_max_permissions creates _MAX_PERMISSIONS_PER_ROLE + 1 permissions (line 365) to test rejection, but there's no test confirming that exactly _MAX_PERMISSIONS_PER_ROLE is accepted. Boundary value testing requires both sides of the limit.



---

**File:** `tests/management/test_mcp_views.py` (line 418)

**low**: create_role_bindings tests don't cover all validation paths - missing test for invalid subject.id while resource.id is valid

test_create_role_bindings_invalid_uuid (line 414) tests invalid role UUID, test_create_role_bindings_invalid_resource_id (line 586) tests invalid resource.id, but there's no test for invalid subject.id. The validation logic checks subject.id at lines 134-139 in mcp_views.py.



---

**File:** `rbac/management/mcp_views.py` (line 100)

**low**: TypeError from non-dict permission silently skipped without reporting to user

At line 100 (diff), when validating V2 permissions, the code checks 'if not isinstance(perm, dict)' and returns an error message. However, this check happens after other validation. If earlier items passed but a later one is not a dict, it returns early. More importantly, the error message is generic and doesn't indicate the actual type that was provided.



---

**File:** `tests/management/test_mcp_views.py` (line 457)

**low**: bulk_delete_roles test checks the first invalid UUID but doesn't verify it catches invalid UUIDs at different positions in the array

test_bulk_delete_roles_invalid_uuid (line 452) passes ['valid-looking', 'not-uuid'] and checks for 'ids[0]' in the error, but 'valid-looking' is also not a valid UUID. The test doesn't verify error reporting for invalid UUIDs at positions other than index 0.



---

**File:** `rbac/management/mcp_views.py` (line 194)

**info**: Write payload validation executes on every write tool call, adding validation overhead to the request path

_validate_write_payload() is called for every write operation after schema validation (line 194-197). For tools like create_role_bindings with many bindings, this iterates through all bindings/roles/UUIDs calling is_valid_uuid() for each, which internally constructs UUID objects via UUID(str(value)) that are immediately discarded.



---

**File:** `rbac/management/mcp_views.py` (line 1)

**informational**: Eight write tools have no semantic validation rules

15 of 23 write tools have explicit validation. The 8 uncovered tools (add_principals_to_group, create_cross_account_request, create_group, create_workspace, delete_group, remove_principals_from_group, remove_roles_from_group, update_group) return None (pass). Most lack UUID params needing validation, so this is defensible.

**Review by EvanCasey13 — APPROVED** [review]
